rfcs/0182-user-management.md

```diff
@@ -0,0 +1,350 @@
+---
+feature: I'm Gonna Build My Own Home Tool With Blackjack and Modules!
+start-date: 2024-09-20
+author: Anderson Torres
+co-authors: `@nyabinary`
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Enhance Nixpkgs monorepo with a declarative management system for basic users.
+
+# Terminology
+[terminology]: #terminology
+
+Henceforth,
+
+- The typical unprivileged user of a system will be called _basic user_;
+- The typical privileged user of a system will be called _superuser_;
+
+# Motivation
+[motivation]: #motivation
+
+[Nixpkgs](https://github.com/NixOS/nixpkgs) is by a far margin the largest
+packageset in this existence, according to
+[Repology](https://repology.org/repository/nix_unstable).
+
+In principle, currently the Nix interpreter and evaluator already leverage
+Nixpkgs for basic users. However, this raw usage is not very ergonomical,
+especially when compared to the structured model provided by NixOS.
+
+In this panorama, the average user has two extreme options here:
+
+- system-wide configuration via NixOS, requiring superuser privileges;
+- ad-hoc management via raw Nix commands, especially the not-recommended
+  `nix-env`.
+
+This RFC proposes to fill this gap: a structured model of system management for
+basic users that does not require superuser privileges to be deployed.
+
+Let's call it `hometool`.
+
+# Detailed Design
+[design]: #detailed-design
+
+`hometool` has two components:
+
+1. A driver program, hereinafter called `hometool`.
+
+   Among other related tasks, this tool has the role of realizing the
+   description of a home environment.
+
+   As expected from a driver, this tool will rely on other programs to execute
+   its intents; the most immediate one being the Nix evaluator.
+
+2. A set of carefully crafted modules, leveraged by Nix Module System.
+
+   The moduleset should reside in `hometool/` directory at the root of Nixpkgs
+   monorepo, similar to how it happens to `nixos/` nowadays.
```

Comment:
There's going to be a lot of shared modules.Should they be placed under `nixos/modules` or their own top-level directory?

I regard this RFC as a reachable milestone of the system-agnostic configuration (#78), which used to be a promising proposal lacking actionable steps. The same framework can be extended to accommodate [nix-darwin](https://github.com/nix-darwin/nix-darwin), [nix-on-droid](https://github.com/nix-community/nix-on-droid), and more. The directory where the shared modules reside could potentially become the API for out-of-tree "configuration projects."

Reply:
I've been switching my config over to [hjem](https://github.com/feel-co/hjem) and [hjem-rum](https://github.com/snugnug/hjem-rum), so I'm leaning more towards having the "core" framework in-tree, and the modules being out-of-tree. So, I can chose to plug in anything I like, as many as I like, or none at all! I like that freedom, so much so that I'd argue that the existing nixos modules should be put out of tree (but that's an orthogonal debate).

But as you said, this could help reach the system-agnostic milestone.